Index: http.h
===================================================================
RCS file: /cvs/src/usr.sbin/httpd/http.h,v
retrieving revision 1.13
diff -u -p -r1.13 http.h
--- http.h	11 Jun 2015 18:49:09 -0000	1.13
+++ http.h	18 Feb 2017 19:48:14 -0000
@@ -242,8 +242,9 @@ struct http_descriptor {
 	int			 http_chunked;
 	char			*http_version;

-	/* Rewritten path remains NULL if not used */
+	/* Rewritten path and query remain NULL if not used */
 	char			*http_path_alias;
+	char			*http_query_alias;

 	/* A tree of headers and attached lists for repeated headers. */
 	struct kv		*http_lastheader;
Index: httpd.conf.5
===================================================================
RCS file: /cvs/src/usr.sbin/httpd/httpd.conf.5,v
retrieving revision 1.73
diff -u -p -r1.73 httpd.conf.5
--- httpd.conf.5	9 May 2016 19:36:54 -0000	1.73
+++ httpd.conf.5	18 Feb 2017 19:48:14 -0000
@@ -383,6 +383,8 @@ The pattern may contain captures that ca
 .Ar uri
 of an enclosed
 .Ic block return
+or
+.Ic pass rewrite
 option.
 .It Oo Ic no Oc Ic log Op Ar option
 Set the specified logging options.
@@ -437,10 +439,23 @@ Enable or disable logging to
 .Xr syslog 3
 instead of the log files.
 .El
-.It Ic pass
+.It Ic pass Op Ic rewrite Ar uri
 Disable any previous
 .Ic block
 in a location.
+If the optional
+.Ar uri
+argument is specified,
+.Xr httpd 8
+will use
+.Ar uri
+instead of the matched location string when determining what
+to load.
+The
+.Ar uri
+may contain predefined macros as explained in the
+.Ic block return
+option.
 .It Ic root Ar option
 Configure the document root and options for the request path.
 Valid options are:
Index: parse.y
===================================================================
RCS file: /cvs/src/usr.sbin/httpd/parse.y,v
retrieving revision 1.78
diff -u -p -r1.78 parse.y
--- parse.y	21 Jun 2016 21:35:24 -0000	1.78
+++ parse.y	18 Feb 2017 19:48:14 -0000
@@ -134,7 +134,7 @@ typedef struct {
 %token	LOCATION LOG LOGDIR MATCH MAXIMUM NO NODELAY ON PORT PREFORK PROTOCOLS
 %token	REQUEST REQUESTS ROOT SACK SERVER SOCKET STRIP STYLE SYSLOG TCP TIMEOUT
 %token	TLS TYPE TYPES HSTS MAXAGE SUBDOMAINS DEFAULT PRELOAD
-%token	ERROR INCLUDE AUTHENTICATE WITH BLOCK DROP RETURN PASS
+%token	ERROR INCLUDE AUTHENTICATE WITH BLOCK DROP RETURN PASS REWRITE
 %token	<v.string>	STRING
 %token  <v.number>	NUMBER
 %type	<v.port>	port
@@ -933,10 +933,12 @@ filter		: block RETURN NUMBER optstring
 			/* Forbidden */
 			srv_conf->return_code = 403;
 		}
-		| PASS				{
-			srv_conf->flags &= ~SRVFLAG_BLOCK;
-			srv_conf->flags |= SRVFLAG_NO_BLOCK;
-		}
+		| pass REWRITE STRING	{
+			free(srv_conf->return_uri);
+			srv_conf->return_uri = $3;
+			srv_conf->return_uri_len = strlen($3) + 1;
+ 		}
+		| pass
 		;

 block		: BLOCK				{
@@ -945,6 +947,12 @@ block		: BLOCK				{
 		}
 		;

+pass		: PASS				{
+			srv_conf->flags &= ~SRVFLAG_BLOCK;
+			srv_conf->flags |= SRVFLAG_NO_BLOCK;
+		}
+		;
+
 optmatch	: /* empty */		{ $$ = 0; }
 		| MATCH			{ $$ = 1; }
 		;
@@ -1192,6 +1200,7 @@ lookup(char *s)
 		{ "request",		REQUEST },
 		{ "requests",		REQUESTS },
 		{ "return",		RETURN },
+		{ "rewrite",		REWRITE },
 		{ "root",		ROOT },
 		{ "sack",		SACK },
 		{ "server",		SERVER },
Index: server_fcgi.c
===================================================================
RCS file: /cvs/src/usr.sbin/httpd/server_fcgi.c,v
retrieving revision 1.68
diff -u -p -r1.68 server_fcgi.c
--- server_fcgi.c	24 Apr 2016 20:09:45 -0000	1.68
+++ server_fcgi.c	18 Feb 2017 19:48:14 -0000
@@ -97,7 +97,7 @@ server_fcgi(struct httpd *env, struct cl
 	int				 pathlen;
 	int				 fd = -1, ret;
 	const char			*stripped, *p, *alias, *errstr = NULL;
-	char				*str, *script = NULL;
+	char				*queryalias, *str, *script = NULL;

 	if (srv_conf->socket[0] == ':') {
 		struct sockaddr_storage	 ss;
@@ -194,6 +194,10 @@ server_fcgi(struct httpd *env, struct cl
 	    ? desc->http_path_alias
 	    : desc->http_path;

+	queryalias = desc->http_query_alias != NULL
+	    ? desc->http_query_alias
+	    : desc->http_query;
+
 	stripped = server_root_strip(alias, srv_conf->strip);
 	if ((pathlen = asprintf(&script, "%s%s", srv_conf->root, stripped))
 	    == -1) {
@@ -242,8 +246,8 @@ server_fcgi(struct httpd *env, struct cl
 		goto fail;
 	}

-	if (desc->http_query) {
-		if (fcgi_add_param(&param, "QUERY_STRING", desc->http_query,
+	if (queryalias) {
+		if (fcgi_add_param(&param, "QUERY_STRING", queryalias,
 		    clt) == -1) {
 			errstr = "failed to encode param";
 			goto fail;
Index: server_http.c
===================================================================
RCS file: /cvs/src/usr.sbin/httpd/server_http.c,v
retrieving revision 1.108
diff -u -p -r1.108 server_http.c
--- server_http.c	27 May 2016 11:24:13 -0000	1.108
+++ server_http.c	18 Feb 2017 19:48:15 -0000
@@ -1046,6 +1046,7 @@ server_response(struct httpd *httpd, str
 {
 	char			 path[PATH_MAX];
 	char			 hostname[HOST_NAME_MAX+1];
+	char			 buf[IBUF_READ_SIZE];
 	struct http_descriptor	*desc = clt->clt_descreq;
 	struct http_descriptor	*resp = clt->clt_descresp;
 	struct server		*srv = clt->clt_srv;
@@ -1163,7 +1164,28 @@ server_response(struct httpd *httpd, str
 		server_abort_http(clt, srv_conf->return_code,
 		    srv_conf->return_uri);
 		return (-1);
-	} else if (srv_conf->flags & SRVFLAG_AUTH &&
+	} else if (srv_conf->return_uri) {
+		memset(buf, 0, sizeof(buf));
+
+		if (server_expand_http(clt, srv_conf->return_uri, buf,
+		    sizeof(buf)) == NULL)
+			goto fail;
+
+		free(desc->http_path_alias);
+		if ((desc->http_path_alias = strdup(buf)) == NULL)
+			goto fail;
+
+		free(desc->http_query_alias);
+		desc->http_query_alias = strchr(desc->http_path_alias, '?');
+		if (desc->http_query_alias != NULL) {
+			*desc->http_query_alias++ = '\0';
+			if ((desc->http_query_alias =
+			    strdup(desc->http_query_alias)) == NULL)
+				goto fail;
+		}
+	}
+
+	if (srv_conf->flags & SRVFLAG_AUTH &&
 	    server_http_authenticate(srv_conf, clt) == -1) {
 		server_abort_http(clt, 401, srv_conf->auth_realm);
 		return (-1);
